packages/flutter/lib/src/widgets/scrollable_grid.dart

```diff
@@ -0,0 +1,219 @@
+
+  void _handleExtentsChanged(double contentExtent, double containerExtent) {
+    setState(() {
+      scrollTo(scrollBehavior.updateExtents(
+        contentExtent: contentExtent,
+        containerExtent: containerExtent,
+        scrollOffset: scrollOffset
+      ));
+    });
+  }
+
+  Widget buildContent(BuildContext context) {
+    return new GridViewport(
+      startOffset: scrollOffset,
+      delegate: config.delegate,
+      onExtentsChanged: _handleExtentsChanged,
+      children: config.children
+    );
+  }
+}
+
+typedef void ExtentsChangedCallback(double contentExtent, double containerExtent);
+
+class GridViewport extends RenderObjectWidget {
+  GridViewport({
+    Key key,
+    this.startOffset,
+    this.delegate,
+    this.onExtentsChanged,
+    this.children
+  });
+
+  final double startOffset;
+  final GridDelegate delegate;
+  final ExtentsChangedCallback onExtentsChanged;
+  final List<Widget> children;
+
+  RenderGrid createRenderObject() => new RenderGrid(delegate: delegate);
+
+  _GridViewportElement createElement() => new _GridViewportElement(this);
+}
+
+// TODO(abarth): This function should go somewhere more general.
+int _lowerBound(List sortedList, var value, { int begin: 0 }) {
+  int current = begin;
+  int count = sortedList.length - current;
+  while (count > 0) {
+    int step = count >> 1;
+    int test = current + step;
+    if (sortedList[test] < value) {
+      current = test + 1;
+      count -= step + 1;
+    } else {
+      count = step;
+    }
+  }
+  return current;
+}
+
+class _GridViewportElement extends RenderObjectElement<GridViewport> {
+  _GridViewportElement(GridViewport widget) : super(widget);
+
+  double _contentExtent;
+  double _containerExtent;
+
+  int _firstMaterializedChildIndex;
+  int _materializedChildCount;
+  List<Element> _materializedChildren = const <Element>[];
+
+  GridSpecification _specification;
+  int _firstMaterializedRowIndex;
+  int _materializedRowLimit;
+
+  RenderGrid get renderObject => super.renderObject;
+
+  void visitChildren(ElementVisitor visitor) {
+    if (_materializedChildren == null)
+      return;
+    for (Element child in _materializedChildren)
+      visitor(child);
+  }
+
+  void mount(Element parent, dynamic newSlot) {
+    super.mount(parent, newSlot);
+    renderObject.callback = layout;
+    _updateRenderObject();
+  }
+
+  void unmount() {
+    renderObject.callback = null;
+    super.unmount();
+  }
+
+  void update(GridViewport newWidget) {
+    super.update(newWidget);
+    _updateRenderObject();
+    if (!renderObject.needsLayout)
+      _updateChildren();
+  }
+
+  void _updatePaintOffset() {
+    renderObject.paintOffset = new Offset(0.0, -(widget.startOffset - _specification.rowOffsets[_firstMaterializedRowIndex]));
+  }
+
+  void _updateRenderObject() {
+    renderObject.delegate = widget.delegate;
+    renderObject.virtualChildCount = widget.children.length;
+
+    if (_specification != null) {
```

Comment:
I'm fine with it either way. I default to if-blocks because the language I grew up with didn't have early return at the time, but there's no particular reason to prefer that to early return in general. (There might be in specific cases, but I don't see one here.)